ref/haskell/src/Lib.hs

```diff
@@ -0,0 +1,114 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+module Lib
+  ( bech32Encode
+  , bech32Decode
+  , segwitEncode
+  , segwitDecode
+  ) where
+
+import Control.Monad (guard)
+import Data.Bits (Bits, unsafeShiftL, unsafeShiftR, (.&.), xor)
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Char8 as BSC
+import Data.Char (toLower, toUpper)
+import Data.Digits
+import Data.Foldable (foldl')
+import qualified Data.HashMap.Strict as Map
+import Data.Word (Word8)
+
+type HRP = BS.ByteString
+type Data = [Word8]
+
+(.>>.), (.<<.) :: Bits a => a -> Int -> a
+(.>>.) = unsafeShiftR
+(.<<.) = unsafeShiftL
+
+charset :: String
+charset = "qpzry9x8gf2tvdw0s3jn54khce6mua7l"
+
+charsetMap :: Map.HashMap Char Word8
+charsetMap = Map.fromList $ zip charset [0..]
+
+charsetMapInv :: Map.HashMap Word8 Char
+charsetMapInv = Map.fromList $ zip [0..] charset
+
+bech32Polymod :: Data -> Int
+bech32Polymod = foldl' go 1
+  where
+    go chk value = foldl' xor chk' [(- (top .>>. i) .&. 1) .&. g | (g, i) <- zip generator [0..]]
+      where
+        generator = [0x3b6a57b2, 0x26508e6d, 0x1ea119fa, 0x3d4233dd, 0x2a1462b3]
+        top = chk .>>. 25
+        chk' = (chk .&. 0x1ffffff) .<<. 5 `xor` (fromIntegral value)
```

Comment:
As far as I can tell, `Word32` doesn't matter (if the type is bigger, it would just fill the higher bits and discard them later).

Even so, I tried your suggestion, and weirdly, the unit tests fail:

```
Addresses:         FAIL
    "BC1QW508D6QEJXTDG4Y5R3ZARVARY0C5XW7KV8F3T4"
    expected: Just "bc1qw508d6qejxtdg4y5r3zarvary0c5xw7kv8f3t4"
     but got: Just "bc1qw508d6qejxtdg4y5r3zarvary0c5xw7kq8f3t4"
```

Same error happens in the python ref unit tests when I apply the same change. I don't quite see why yet, seems like it should work.